work-pool/src/WorkPool.hs

```diff
@@ -0,0 +1,119 @@
+module WorkPool
+  ( Config(..)
+  , Done
+  , Pool
+  , Source
+  , pushJob
+  , readJobCount
+  , runPool
+  , runSource
+  )
+where
+
+import Data.Set (Set)
+import MPrelude
+import Prelude (succ)
+
+import qualified Data.Set       as Set
+import qualified UnliftIO.Async as UnliftIO
+import qualified UnliftIO.STM   as UnliftIO
+
+-- | Worker pool configuration
+data Config m a = Config
+  { produceJobs :: Pool a -> m ()
+  -- ^ function called to produce work, use `pushJob` to create workable jobs.
+  -- once this function returns and all jobs are worked off the workers exit.
+  , queueSize   :: Natural
+  -- ^ maximum size of the jobs queued
+  , workerCount :: Natural
+  -- ^ number of workers to boot
+  , workerRun   :: Natural -> Source a -> m Done
+  -- ^ function called when a worker is booted, argument is the worker index,
+  -- and a source to be drained with `runSource`
+  }
+
+-- Internal queue event, supplying job or quitting the worker
+data Event a = Quit | Job a
+
+-- | Running pool
+newtype Pool a = Pool
+  { queue :: UnliftIO.TBQueue (Event a)
+  }
+
+-- | Source to be drained
+newtype Source a = Source
+  { queue :: UnliftIO.TBQueue (Event a)
+  }
+
+-- | Type forcing clients to drain the source via `runSource`
+data Done = Done
+
+-- | Add (dynamically) created a job to the pool
+--
+-- This function will block if the max queue size would be overflown.
+-- As the workers create space in the queue this function will unblock.
+pushJob :: MonadIO m => Pool a -> a -> m ()
+pushJob Pool{..} item
+  = UnliftIO.atomically
+  $ UnliftIO.writeTBQueue queue (Job item)
+
+-- | Read the number of jobs in the pool
+--
+-- This function does not block. Its not guaranteed count is still correct
+-- when the function returns
+readJobCount :: MonadIO m => Pool a -> m Natural
+readJobCount Pool{..}
+  = UnliftIO.atomically
+  $ UnliftIO.lengthTBQueue queue
+
+-- | Drain a source
+--
+-- This function:
+-- * is executed 0 or many times per worker.
+-- * executes the action as long there are items in the queue.
+-- * as long the producer did not exit: blocks if there are no items in the queue
+-- * if the producer exits and the queue is empty: returns.
+-- * is he only way to produce a `Done` value required by the `Config` api.
+runSource :: MonadIO m => (a -> m ()) -> Source a -> m Done
+runSource action Source{..} = go $> Done
+  where
+    go = UnliftIO.atomically (UnliftIO.readTBQueue queue) >>= \case
+      (Job item) -> action item >> go
```

Comment:
__question: (nitpick)__ Are the parenthesis needed here?

Reply:
yes